bcipy/helpers/raw_data.py

```diff
@@ -0,0 +1,303 @@
+        print(f"Data from ${reader.daq_type}")
+        print(reader.columns)
+        for row in reader:
+            print(row)
+    ```
+
+    Parameters
+    ----------
+    - file_path : path to the csv file
+    - convert_data : if True attempts to convert data values to floats;
+    default is False
+    """
+
+    def __init__(self, file_path: str, convert_data: bool = False):
+        self.file_path = file_path
+        self._file_obj = None
+        self.convert_data = convert_data
+
+    def __enter__(self):
+        self._file_obj = open(self.file_path, mode="r")
+        self.daq_type, self.sample_rate = read_metadata(self._file_obj)
+        self._reader = csv.reader(self._file_obj)
+        self.columns = next(self._reader)
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """Exit the context manager. Close resources"""
+        self._file_obj.close()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        assert self._reader, "Reader must be initialized"
+        row = next(self._reader)
+        if self.convert_data:
+            return list(map(maybe_float, row))
+        return row
+
+
+class RawDataWriter:
+    """Writes a raw data file one row at a time without storing the records
+    in memory. Intended to be used as a ContextManager using Python's `with`
+    keyword.
+
+    Example usage:
+
+    ```
+    with RawDataWriter(path, daq_type, sample_rate, columns) as writer:
+        for row in data:
+            writer.writerow(row)
+    ```
+
+    Parameters
+    ----------
+    - file_path : path to the csv file
+    - daq_type : name of device
+    - sample_rate : sample frequency in Hz
+    - columns : list of column names
+    """
+
+    def __init__(self, file_path: str, daq_type: str, sample_rate: float,
+                 columns: List[str]):
+        self.file_path = file_path
+        self.daq_type = daq_type
+        self.sample_rate = sample_rate
+        self.columns = columns
+        self._file_obj = None
+
+    def __enter__(self):
+        self._file_obj = open(self.file_path,
+                              mode='w',
+                              encoding='utf-8',
+                              newline='')
+        # write metadata
+        self._file_obj.write(f"daq_type,{self.daq_type}\n")
+        self._file_obj.write(f"sample_rate,{self.sample_rate}\n")
+
+        # If flush is missing the previous content may be appended at the end.
+        self._file_obj.flush()
+
+        self._csv_writer = csv.writer(self._file_obj, delimiter=',')
+        self._csv_writer.writerow(self.columns)
+
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """Exit the context manager. Close resources"""
+        self._file_obj.close()
+
+    def writerow(self, row: List) -> None:
+        assert self._csv_writer, "Writer must be initialized"
+        self._csv_writer.writerow(row)
+
+
+def load(filename: str) -> RawData:
```

Comment:
I like that we can support both full read and iterative. It makes a lot of sense